fclient/performjoin.go

```diff
@@ -0,0 +1,238 @@
+	respMakeJoin.JoinEvent.Sender = input.UserID.Raw()
+	respMakeJoin.JoinEvent.StateKey = &stateKey
+	respMakeJoin.JoinEvent.RoomID = input.RoomID
+	respMakeJoin.JoinEvent.Redacts = ""
+	if input.Content == nil {
+		input.Content = map[string]interface{}{}
+	}
+	_ = json.Unmarshal(respMakeJoin.JoinEvent.Content, &input.Content)
+	input.Content["membership"] = spec.Join
+	if err = respMakeJoin.JoinEvent.SetContent(input.Content); err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.SetContent: %w", err)
+	}
+	if err = respMakeJoin.JoinEvent.SetUnsigned(struct{}{}); err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.SetUnsigned: %w", err)
+	}
+
+	// Work out if we support the room version that has been supplied in
+	// the make_join response.
+	// "If not provided, the room version is assumed to be either "1" or "2"."
+	// https://matrix.org/docs/spec/server_server/unstable#get-matrix-federation-v1-make-join-roomid-userid
+	if respMakeJoin.RoomVersion == "" {
+		respMakeJoin.RoomVersion = setDefaultRoomVersionFromJoinEvent(respMakeJoin.JoinEvent)
+	}
+	verImpl, err := gomatrixserverlib.GetRoomVersion(respMakeJoin.RoomVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Build the join event.
+	event, err := respMakeJoin.JoinEvent.Build(
+		time.Now(),
+		origin,
+		input.KeyID,
+		input.PrivateKey,
+		respMakeJoin.RoomVersion,
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("respMakeJoin.JoinEvent.Build: %w", err)
+	}
+
+	var respState gomatrixserverlib.StateResponse
+	// Try to perform a send_join using the newly built event.
+	respSendJoin, err := fedClient.SendJoin(
+		context.Background(),
+		origin,
+		input.ServerName,
+		event,
+	)
+	if err != nil {
+		callbacks.FederationFailure(input.ServerName)
+		return nil, nil, fmt.Errorf("r.federation.SendJoin: %w", err)
+	}
+	callbacks.FederationSuccess(input.ServerName)
+
+	// If the remote server returned an event in the "event" key of
+	// the send_join request then we should use that instead. It may
+	// contain signatures that we don't know about.
+	if len(respSendJoin.Event) > 0 {
+		var remoteEvent *gomatrixserverlib.Event
+		remoteEvent, err = verImpl.NewEventFromUntrustedJSON(respSendJoin.Event)
+		if err == nil && isWellFormedMembershipEvent(
+			remoteEvent, input.RoomID, input.UserID,
+		) {
+			event = remoteEvent
+		}
+	}
+
+	// Sanity-check the join response to ensure that it has a create
+	// event, that the room version is known, etc.
+	authEvents := respSendJoin.AuthEvents.UntrustedEvents(respMakeJoin.RoomVersion)
+	if err = checkEventsContainCreateEvent(authEvents); err != nil {
+		return nil, nil, fmt.Errorf("sanityCheckAuthChain: %w", err)
+	}
+
+	// Process the join response in a goroutine. The idea here is
+	// that we'll try and wait for as long as possible for the work
+	// to complete, but if the client does give up waiting, we'll
+	// still continue to process the join anyway so that we don't
+	// waste the effort.
```

Comment:
The key thing that does this is using a background context rather than the provided `ctx`.